crates/subspace-farmer/src/utils/archival_storage_info.rs

```diff
@@ -0,0 +1,125 @@
+}
+
+struct PeerFilter {
+    cuckoo_filter: CuckooFilter<DefaultHasher>,
+    connected: bool,
+}
+
+impl Debug for ArchivalStorageInfo {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("ArchivalStorageInfo")
+            .field("peers (len)", &self.peers.lock().len())
+            .finish()
+    }
+}
+
+impl ArchivalStorageInfo {
+    pub fn update_cuckoo_filter(
+        &mut self,
+        peer_id: PeerId,
+        cuckoo_filter_dto: Arc<CuckooFilterDTO>,
+        currently_connected_peers: &[PeerId],
+    ) {
+        let currently_connected_peers =
+            HashSet::<PeerId>::from_iter(currently_connected_peers.iter().cloned());
+
+        let exported_filter = ExportedCuckooFilter {
+            values: cuckoo_filter_dto.values.clone(),
+            length: cuckoo_filter_dto.length as usize,
+        };
+
+        let cuckoo_filter = CuckooFilter::from(exported_filter);
+
+        let mut peer_filters = self.peers.lock();
+
+        peer_filters.insert(
+            peer_id,
+            PeerFilter {
+                cuckoo_filter,
+                connected: true,
+            },
+        );
+
+        let mut connected_peers = Vec::new();
+        let mut disconnected_peers = Vec::new();
+
+        for (peer_id, peer_filter) in peer_filters.iter_mut() {
+            let connected = currently_connected_peers.contains(peer_id);
+            peer_filter.connected = connected;
+
+            if connected {
+                connected_peers.push(*peer_id);
+            } else {
+                disconnected_peers.push(*peer_id);
+            }
+        }
+
+        // Truncate current peer set by limits.
+        let exceeding_number_of_connected_peers = connected_peers
+            .len()
+            .saturating_div(CONNECTED_PEERS_NUMBER_LIMIT);
+        let exceeding_number_of_disonnected_peers = connected_peers
+            .len()
+            .saturating_div(DISCONNECTED_PEERS_NUMBER_LIMIT);
+
+        let mut rng = StdRng::seed_from_u64({
+            // Hash of PeerID
+            let mut s = DefaultHasher::new();
+            peer_id.hash(&mut s);
+            s.finish()
+        });
+
+        for _ in 0..exceeding_number_of_connected_peers {
+            let random_index = rng.gen_range(0..connected_peers.len());
+
+            let removing_peer_id = *connected_peers
+                .get(random_index)
+                .expect("Index is checked to be present.");
+
+            connected_peers.remove(random_index);
+            peer_filters.remove(&removing_peer_id);
+        }
+
+        for _ in 0..exceeding_number_of_disonnected_peers {
+            let random_index = rng.gen_range(0..disconnected_peers.len());
+
+            let removing_peer_id = *disconnected_peers
+                .get(random_index)
+                .expect("Index is checked to be present.");
+
+            disconnected_peers.remove(random_index);
+            peer_filters.remove(&removing_peer_id);
+        }
```

Comment:
What is all of this? There is no description and I don't understand what is going on here. Also I really don't like two more constants that are hardcoded in implementation and need to be kept in sync with the rest of the codebase.